File: packages/core/src/routes/notes.ts
/**
 * Note routes — thin HTTP layer over NoteService.
 */

import { Hono } from 'hono'
import type { Services } from '../services'
import { mapStatusCode } from './server'

export function noteRoutes(services: Services) {
  const app = new Hono()

  app.post('/notes', async (c) => {
    const body = await c.req.json<{ title?: string; content: string }>()
    const result = await services.notes.create(body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/notes', async (c) => {
    const offset = Math.max(0, parseInt(c.req.query('offset') ?? '0', 10) || 0)
    const limit = Math.min(200, Math.max(1, parseInt(c.req.query('limit') ?? '50', 10) || 50))
    const search = c.req.query('search')

    const result = await services.notes.list(search, offset, limit)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data, pagination: result.pagination })
  })

  app.get('/notes/:id', async (c) => {
    const result = await services.notes.get(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/notes/:id', async (c) => {
    const body = await c.req.json<{ title?: string; content?: string }>()
    const result = await services.notes.update(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/notes/:id', async (c) => {
    const result = await services.notes.delete(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  app.get('/notes/by-entity/:entityType/:entityId', async (c) => {
    const result = await services.notes.getNotesForEntity(
      c.req.param('entityType'),
      c.req.param('entityId'),
    )
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.post('/notes/:id/references', async (c) => {
    const body = await c.req.json<{ entity_type: string; entity_id: string }>()
    const result = await services.notes.addReference(c.req.param('id'), body.entity_type, body.entity_id)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: null }, 201)
  })

  app.delete('/notes/:id/references/:entityType/:entityId', async (c) => {
    const result = await services.notes.removeReference(
      c.req.param('id'),
      c.req.param('entityType'),
      c.req.param('entityId'),
    )
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  return app
}

File: crates/forge-server/src/routes/notes.rs
//! Note CRUD routes.
//!
//! Mirrors `packages/core/src/routes/notes.ts` — same URL paths,
//! same JSON shapes, so the webui and MCP server continue working.

use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;

use forge_core::UserNote;
use forge_sdk::db::NoteStore;

use crate::db::with_conn;
use crate::error::ApiError;
use crate::response::{ApiData, ApiList, Created, NoContent};
use crate::state::SharedState;

// ── Query params ────────────────────────────────────────────────────

#[derive(Debug, Deserialize, Default)]
pub struct NoteListQuery {
    pub offset: Option<i64>,
    pub limit: Option<i64>,
    pub search: Option<String>,
}

// ── Request bodies ──────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct CreateNoteBody {
    pub title: Option<String>,
    pub content: String,
}

#[derive(Debug, Deserialize)]
pub struct UpdateNoteBody {
    pub title: Option<String>,
    pub content: Option<String>,
}

// ── Handlers ────────────────────────────────────────────────────────

async fn create_note(
    State(state): State<SharedState>,
    Json(body): Json<CreateNoteBody>,
) -> Result<Created<UserNote>, ApiError> {
    let result = with_conn(&state, move |conn| {
        NoteStore::create(conn, body.title.as_deref(), &body.content)
    })
    .await?;
    Ok(Created(result))
}

async fn list_notes(
    State(state): State<SharedState>,
    Query(q): Query<NoteListQuery>,
) -> Result<Json<ApiList<UserNote>>, ApiError> {
    let offset = q.offset.unwrap_or(0).max(0);
    let limit = q.limit.unwrap_or(50).clamp(1, 200);
    let search = q.search;

    let (data, pagination) = with_conn(&state, move |conn| {
        NoteStore::list(conn, search.as_deref(), offset, limit)
    })
    .await?;

    Ok(Json(ApiList { data, pagination }))
}

async fn get_note(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<UserNote>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        NoteStore::get(conn, &id)?.ok_or_else(|| forge_core::ForgeError::NotFound {
            entity_type: "Note".into(),
            id: id.clone(),
        })
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn update_note(
    State(state): State<SharedState>,
    Path(id): Path<String>,
    Json(body): Json<UpdateNoteBody>,
) -> Result<Json<ApiData<UserNote>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        NoteStore::update(conn, &id, body.title.as_deref(), body.content.as_deref())
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn delete_note(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| NoteStore::delete(conn, &id)).await?;
    Ok(NoContent)
}

// ── Router ──────────────────────────────────────────────────────────

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/notes", post(create_note).get(list_notes))
        .route(
            "/notes/{id}",
            get(get_note).patch(update_note).delete(delete_note),
        )
}
